### PR TITLE
Fix objects using shaders when there is a lidar in the scene (ogre2)

### PR DESCRIPTION
**Pull request by iche033**
Signed-off-by: Ian Chen <ichen@osrfoundation.org>

# 🦟 Bug fix

## Summary

The gpu ray sensor sets up generates retro values by switching materials. The probably is that it incorrectly restores shader materials causing objects using custom shaders to render incorrectly after first update.

Applied a similar fix as as https://github.com/ignitionrobotics/ign-rendering/pull/533. Updated integration test should which fail without this fix.

On way to reproduce the issue is to run this [shader_param_gpu_lidar.sdf](https://gist.github.com/iche033/e74b66edcf4a95554b6a015c72cba2b7) world with ign-gazebo:

```
ign gazebo -v 4 -r shader_param_gpu_lidar.sdf
```

Without this fix, the red sphere and waves (which use custom shaders) will appear black in the `Image Display` plugin.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: ogre2/src/Ogre2GpuRays.cc [hunk 1]
```diff
@@ -324,12 +337,24 @@ void Ogre2LaserRetroMaterialSwitcher::cameraPostRenderScene(
     subItem->setDatablock(it.second);
   }
 
+  auto lIt = laserRetroMaterialMap.find(this);
```
Comment (ahcorde) [inline ogre2/src/Ogre2GpuRays.cc:340]: use un nombre de variable más descriptivo

File: ogre2/src/Ogre2MaterialSwitcher.cc [hunk 2]
```diff
@@ -110,6 +110,13 @@ void Ogre2MaterialSwitcher::cameraPreRenderScene(
       {
         materialMap[this][subItem] = subItem->getMaterial();
         subItem->setMaterial(this->plainMaterial);
+        auto technique = subItem->getMaterial()->getTechnique(0);
+
+        if (technique && !technique->isDepthWriteEnabled() &&
+            !technique->isDepthCheckEnabled())
+          subItem->setMaterial(this->plainOverlayMaterial);
+        else
+          subItem->setMaterial(this->plainMaterial);
```
Comment (ahcorde) [inline ogre2/src/Ogre2MaterialSwitcher.cc:119]:
```suggestion
        if (technique && !technique->isDepthWriteEnabled() &&
            !technique->isDepthCheckEnabled())
        {
          subItem->setMaterial(this->plainOverlayMaterial);
        }
        else
        {
          subItem->setMaterial(this->plainMaterial);
        }
```

**Review by ahcorde — APPROVED** [review]
I made two style suggestions, I tested locally and it works great

File: ogre2/src/Ogre2GpuRays.cc (hunk 1 above)
Comment (iche033) [reply ogre2/src/Ogre2GpuRays.cc]: está hecho [a6c13de](https://github.com/ignitionrobotics/ign-rendering/pull/575/commits/a6c13de99012a85048edbd411959cd36cfff932c)

File: ogre2/src/Ogre2MaterialSwitcher.cc (hunk 2 above)
Comment (iche033) [reply ogre2/src/Ogre2MaterialSwitcher.cc]: [a6c13de](https://github.com/ignitionrobotics/ign-rendering/pull/575/commits/a6c13de99012a85048edbd411959cd36cfff932c)

**Pull request merged**